File: src/languages/english-in/english-in.ts
/**
 * Indian English (en-IN) — "General Indian English" (GIE), an ACCENT VARIANT of the General-American `en` engine
 * (not a separate language). Reuses the full English G2P (dict + heteronyms + OOV model) and applies a phonological
 * DELTA — a context-free surface remap — to the GenAm output, like en-GB/es-419 (no information is lost, so this is
 * a post-process, not an engine fork). Labels the OmniVoice FLEURS `en_in` training audio (~200M L2 speakers).
 *
 * The delta (GenAm → GIE), from the documented phonology (Wells 1982 vol. 3; Sailaja 2009, *Indian English*):
 *   • RETROFLEXION — the signature: alveolar /t d/ → retroflex [ʈ ɖ] (tin→ʈɪn, dog→ɖɔːɡ). Guarded so the tied
 *     affricates t͡ʃ/d͡ʒ (church, judge) are NOT retroflexed.
 *   • TH-STOPPING — the dental fricatives become dental STOPS: /θ/ → [t̪ʰ] (aspirated), /ð/ → [d̪] (thin→t̪ʰɪn,
 *     this→d̪ɪs). Distinct from the retroflex /t d/ by PLACE (dental vs retroflex), so "thin"≠"tin".
 *   • /v/–/w/ MERGER → the labiodental approximant [ʋ] (wet=vet=ʋɛʈ).
 *   • MONOPHTHONGISATION — FACE [eᶦ]→[eː], GOAT [oᶷ]→[oː] (they→d̪eː, goat→ɡoːʈ). PRICE/MOUTH stay diphthongs.
 *   • DE-ASPIRATION of the other voiceless stops /p k/ → [p k] (GIE lacks English aspiration; cat→kæʈ). /t/'s
 *     aspiration is already removed by the retroflexion.
 *   • CLEAR /l/ in all positions (no dark [ɫ]): full→fʊl.
 *   • RHOTIC with a TAP: GIE is rhotic (spelling-pronunciation) — coda /ɹ/ is KEPT and every /ɹ/ → [ɾ]; the
 *     r-coloured vowels de-rhoticise to a plain vowel + [ɾ] (NURSE ɝ→əɾ, lettER ɚ→əɾ; car→kɑːɾ, letter→lɛʈəɾ).
 *   • Reduced [ᵻ]→[ɪ] (GIE is syllable-timed with fuller vowels; the further full-vowel restoration of unstressed
 *     syllables needs the spelling → deferred). The palatal on-glide [ʲ] is dropped.
 *
 * Deliberately DEFERRED (need a lexical distinction / not recoverable from the GenAm output, and no en-IN referee
 * exists to adjudicate): the TRAP/DRESS ([æ]~[ɛ]) and LOT/THOUGHT vowel qualities, yod-retention (student→stjuːɖənʈ),
 * and the syllable-timed full-vowel restoration in unstressed syllables. Quality anchor = a hand-adjudicated
 * diagnostic gold (english-in.test.ts).
 */
import { createEnglish, type EnglishPhonemizer } from "../english/english.ts";

const TIE = "͡"; // U+0361 combining double inverted breve (affricate tie) — guard so t͡ʃ/d͡ʒ aren't retroflexed

/** GenAm citation IPA → General Indian English (context-free; no lexical sets needed). */
export function toIndian(genAm: string): string {
    let s = genAm;
    // RETROFLEXION of /t d/ → [ʈ ɖ], guarded against the tied affricates t͡ʃ/d͡ʒ. Matches t/d + an optional
    // aspiration/flap diacritic, but NOT when the tie U+0361 follows (that is an affricate). Runs BEFORE TH-stopping
    // so the dental stops [t̪ d̪] created below are not swept up.
    s = s.replace(new RegExp(`t[ʰ̬̰]?(?!${TIE})`, "gu"), "ʈ");
    s = s.replace(new RegExp(`d[̬̰]?(?!${TIE})`, "gu"), "ɖ");
    // TH-STOPPING → dental stops (after retroflexion; distinct from ʈ/ɖ by place).
    s = s.replace(/θ/gu, "t̪ʰ").replace(/ð/gu, "d̪");
    // MONOPHTHONGISATION of FACE/GOAT; other offglides → plain [ɪ]/[ʊ] (PRICE/MOUTH/CHOICE stay diphthongs).
    s = s.replace(/eᶦ/gu, "eː").replace(/oᶷ/gu, "oː");
    s = s.replace(/aᶦ/gu, "aɪ").replace(/aᶷ/gu, "aʊ").replace(/ᶦ/gu, "ɪ").replace(/ᶷ/gu, "ʊ");
    // /v/–/w/ MERGER → [ʋ].
    s = s.replace(/[vw]/gu, "ʋ");
    // DE-ASPIRATION of the remaining voiceless stops (/t/ already de-aspirated via retroflexion).
    s = s.replace(/pʰ/gu, "p").replace(/kʰ/gu, "k");
    // CLEAR /l/ (drop the dark-coda diacritic).
    s = s.replace(/ɫ/gu, "l");
    // RHOTIC with a tap: de-rhoticise the r-coloured vowels to V+[ɾ], then every /ɹ/ → [ɾ] (coda kept — GIE is rhotic).
    s = s.replace(/ɝ/gu, "əɾ").replace(/ɚ/gu, "əɾ").replace(/ɹ/gu, "ɾ");
    // Fuller vowels: reduced [ᵻ]/[ᵿ] → [ɪ]/[ʊ]; drop the palatal on-glide [ʲ].
    s = s.replace(/ᵻ/gu, "ɪ").replace(/ᵿ/gu, "ʊ").replace(/ʲ/gu, "");
    return s;
}

let IN: EnglishPhonemizer | undefined;
const eng = (): EnglishPhonemizer => (IN ??= createEnglish());

/** One en-IN word → canonical IPA (GenAm engine + the GIE delta). The delta is context-free, so the shipped and
 *  rule paths are identical (no mined lexical sets) — the eval, if any, is trivially non-circular. */
export function phonemizeWord(word: string): string {
    return toIndian(eng().text(word));
}
/** Alias kept for parity with the other accent variants (en-GB/pt-BR) that split a lexicon-bearing shipped path. */
export const phonemizeWordRules = phonemizeWord;

/** Build the Indian-English phonemizer (GenAm engine + the GIE delta on each word's output). */
export function createEnglishIN(): { text(input: string): string } {
    const e = createEnglish();
    return { text: (input: string): string => e.text(input, (ipa) => toIndian(ipa)) };
}

File: csharp/Vernacula.Phonemizer/Languages/EnglishIn/EnglishIn.cs
/**
 * Indian English (en-IN) — "General Indian English" (GIE), an ACCENT VARIANT of the General-American `en`
 * engine (not a separate language). Reuses the full English G2P (dict + heteronyms + OOV model) and applies a
 * context-free surface DELTA to its output.
 * Ported from src/languages/english-in/english-in.ts — see that file for the phonological sources
 * (Wells 1982 vol. 3; Sailaja 2009) and for what is deliberately DEFERRED.
 *
 * ⚠ THE DELTA IS APPLIED PER WORD, NOT TO THE WHOLE UTTERANCE, and that is a real difference from es-419 and
 * fr-CA, which wrap the assembled output. English's `Text` takes a `wordTransform`, and en-IN threads the
 * remap through it — so the transform sees one word's IPA at a time. Wrapping the utterance instead would
 * expose the delta to clause punctuation and to any inter-word material the assembler adds.
 *
 * ⚠ NO ROMAN POLICY: `en-IN` is in `Registry.ROMAN_NATIVE`, so the shared Roman pass is skipped and English
 * resolves numerals in its own normalization.
 */
using Vernacula.Phonemizer.Core;
using Vernacula.Phonemizer.Languages.English;

namespace Vernacula.Phonemizer.Languages.EnglishIn;

public static class EnglishIn
{
    // U+0361 COMBINING DOUBLE INVERTED BREVE — the affricate tie. The guard that keeps t͡ʃ / d͡ʒ (church,
    // judge) out of the retroflexion.
    private const string TIE = "͡";

    // ⚠ THE OPTIONAL DIACRITICS ARE INVISIBLE IN A CHARACTER CLASS, so they are written as escapes:
    // U+02B0 MODIFIER LETTER SMALL H (aspiration), U+032C COMBINING CARON BELOW (voiced), U+0330 COMBINING
    // TILDE BELOW (creaky). Written as literals in the TS, where they render as an empty-looking class.
    private static readonly JsRe RETRO_T = JsRegex.Compile($"t[ʰ̬̰]?(?!{TIE})", "gu");
    private static readonly JsRe RETRO_D = JsRegex.Compile($"d[̬̰]?(?!{TIE})", "gu");

    /** GenAm citation IPA → General Indian English (context-free; no lexical sets needed). */
    public static string ToIndian(string genAm)
    {
        // RETROFLEXION of /t d/ → [ʈ ɖ], guarded against the tied affricates. ⚠ RUNS BEFORE TH-STOPPING so the
        // dental stops [t̪ d̪] created below are not swept up by it.
        var s = RETRO_D.Replace(RETRO_T.Replace(genAm, "ʈ"), "ɖ");
        // TH-STOPPING → dental stops (after retroflexion; distinct from ʈ/ɖ by PLACE, so "thin" ≠ "tin").
        s = s.Replace("θ", "t̪ʰ").Replace("ð", "d̪");
        // MONOPHTHONGISATION of FACE/GOAT; other offglides → plain [ɪ]/[ʊ] (PRICE/MOUTH/CHOICE stay diphthongs).
        s = s.Replace("eᶦ", "eː").Replace("oᶷ", "oː");
        s = s.Replace("aᶦ", "aɪ").Replace("aᶷ", "aʊ").Replace("ᶦ", "ɪ").Replace("ᶷ", "ʊ");
        // /v/–/w/ MERGER → [ʋ].
        s = s.Replace("v", "ʋ").Replace("w", "ʋ");
        // DE-ASPIRATION of the remaining voiceless stops (/t/ already de-aspirated via retroflexion).
        s = s.Replace("pʰ", "p").Replace("kʰ", "k");
        // CLEAR /l/ (drop the dark-coda diacritic).
        s = s.Replace("ɫ", "l");
        // RHOTIC WITH A TAP: de-rhoticise the r-coloured vowels to V+[ɾ], then every /ɹ/ → [ɾ]. The coda is
        // KEPT — GIE is rhotic, unlike en-GB.
        s = s.Replace("ɝ", "əɾ").Replace("ɚ", "əɾ").Replace("ɹ", "ɾ");
        // Fuller vowels: reduced [ᵻ]/[ᵿ] → [ɪ]/[ʊ]; drop the palatal on-glide [ʲ].
        return s.Replace("ᵻ", "ɪ").Replace("ᵿ", "ʊ").Replace("ʲ", "");
    }

    private static EnglishPhonemizer? IN;
    private static EnglishPhonemizer Eng() => IN ??= EnglishFactory.CreateEnglish();

    /** One en-IN word → canonical IPA. Context-free, so the shipped path IS the rule path. */
    public static string PhonemizeWord(string word) => ToIndian(Eng().Text(word));

    /** Alias kept for parity with the variants (en-GB/pt-BR) that split a lexicon-bearing shipped path. */
    public static string PhonemizeWordRules(string word) => PhonemizeWord(word);

    private sealed class EnInLanguage(EnglishPhonemizer inner) : ILanguage
    {
        public string Text(string input) => inner.Text(input, (ipa, _) => ToIndian(ipa), null);
    }

    /** Build the Indian-English phonemizer (GenAm engine + the GIE delta on each WORD's output). */
    public static ILanguage CreateEnglishIN() => new EnInLanguage(EnglishFactory.CreateEnglish());

    internal static void RegisterSelf() => Registry.Register("english-in", CreateEnglishIN);
}
